service/parserprovider/expand.go

```diff
@@ -0,0 +1,89 @@
+// Copyright The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parserprovider
+
+import (
+	"context"
+	"os"
+
+	"go.opentelemetry.io/collector/config"
+)
+
+type expandMapProvider struct {
+	base MapProvider
+}
+
+// NewExpandMapProvider returns a MapProvider, that expands all environment variables for a
+// config.Map provided by the given MapProvider.
+func NewExpandMapProvider(base MapProvider) MapProvider {
+	return &expandMapProvider{
+		base: base,
+	}
+}
+
+func (emp *expandMapProvider) Get(ctx context.Context) (*config.Map, error) {
+	cfgMap, err := emp.base.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	for _, k := range cfgMap.AllKeys() {
+		cfgMap.Set(k, expandStringValues(cfgMap.Get(k)))
+	}
+	return cfgMap, nil
+}
+
+func (emp *expandMapProvider) Close(ctx context.Context) error {
+	return emp.base.Close(ctx)
+}
+
+func expandStringValues(value interface{}) interface{} {
+	switch v := value.(type) {
+	default:
+		return v
+	case string:
+		return expandEnv(v)
+	case []interface{}:
+		nslice := make([]interface{}, 0, len(v))
+		for _, vint := range v {
+			nslice = append(nslice, expandStringValues(vint))
+		}
+		return nslice
+	case map[string]interface{}:
+		nmap := make(map[interface{}]interface{}, len(v))
+		for k, vint := range v {
+			nmap[k] = expandStringValues(vint)
+		}
+		return nmap
+	case map[interface{}]interface{}:
```

Comment:
Are you sure that we still hit this case? IIRC after we switched to koanf this case may not be hit.

Reply:
Not sure, but trying to limit the changes in one PR